base/essentials.jl

```diff
@@ -700,14 +701,43 @@ end
 ```
 
 !!! warning
-
     Using `@inbounds` may return incorrect results/crashes/corruption
     for out-of-bounds indices. The user is responsible for checking it manually.
     Only use `@inbounds` when it is certain from the information locally available
-    that all accesses are in bounds. In particular, using `1:length(A)` instead of
-    `eachindex(A)` in a function like the one above is _not_ safely inbounds because
+    that all accesses are in-bounds. In particular, using `1:length(A)` instead of
+    `eachindex(A)` in a function like the one above is _not_ safely in-bounds because
     the first index of `A` may not be `1` for all user defined types that subtype
     `AbstractArray`.
+
+!!! note
+    `@inbounds` eleminates bounds checks that are syntactically within the given block,
+    as well as ones in methods that are called within the block.
+    However, keep in mind that the `@inbounds` context propagates only one function call
+    layer deep. For example, if an `@inbounds` block includes a call to `f()`, which in turn
+    calls `g()`, bounds checks that are syntactically within `f()` will be eliminated,
+    while ones within `g()` will not.
+    If you want to eliminates bounds checks within `g()` also,
+    you need to annotate [`Base.@propagate_inbounds`](@ref) on `f()`.
+
+# Example
+
+```julia-repl
+julia> code_typed((Vector{Any},Int)) do a, i
```

Comment:
The `do` form of `code_typed` is hard to parse for me, I think it is better to extract that part into it's own function. Also, I feel that `arrayref` has to be explicitly commented on and explained that the first argument is if it should be boundschecked.

Reply:
It is also not accurate anymore, since `arrayref` is changed to `memoryrefget` and `Expr(:boundscheck)` is no longer removed, though I am fine with the do-block form. Perhaps this should show the IR for `@boundscheck return true; return false`